### PR TITLE
feat(ddd): create initial implementation

### PR DESCRIPTION
**Pull request by avivharuzi**
## PR Checklist

Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: [CONTRIBUTING.md](CONTRIBUTING.md#commit-message-format)
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type

What kind of change does this PR introduce?

```
[ ] Bugfix
[x] Feature
[ ] Code style update (formatting, local variables)
[x] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[x] Documentation content changes
[ ] Other... Please describe:
```

## What is the current behavior?

Issue Number: N/A

## What is the new behavior?

Updated ddd package with new logic before that it was empty with no usage at all.

The package can generate ddd libraries for Angular or React in the Nx workspace!

I created this library because I found that I repeat myself doing the same library with domain-driven design structure like suggested in Nx documentation to best work with.
There is a good package by Manfred `@angular-architects/ddd` which also generates ddd libraries but the difference that I found was that it only works with Angular... and it generates more things that I found more complex.

So this ddd package will bring a simple solution for people who want to implement a domain-driven design with best practices that works also with both Angular and React libraries and can be extended in the future to other frameworks or libraries.

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

### CODE REVIEW

File: packages/ddd/src/utils/create-angular-component.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+  export: boolean;
+  flat: boolean;
+  prefix: string;
+}
+
+export const createAngularComponent = async (
+  tree: Tree,
+  options: createAngularComponentOptions
+): Promise<void> => {
+  const libraryGenerator = wrapAngularDevkitSchematic(
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/src/utils/create-angular-component.ts:18]: I disagree about it.
Nx recommends using their libraries generators.
https://nx.dev/l/r/generators/workspace-generators

File: packages/ddd/src/generators/react/generator.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
+import { formatFiles, Tree } from '@nrwl/devkit';
+import { Linter } from '@nrwl/linter';
+import { libraryGenerator } from '@nrwl/react';
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/src/generators/react/generator.ts:3]: I disagree about it.
Nx recommends using their libraries generators.
https://nx.dev/l/r/generators/workspace-generators

File: packages/ddd/src/utils/create-angular-component-stories.ts [hunk 3]
```diff
@@ -0,0 +1,11 @@
+import { angularStoriesGenerator } from '@nrwl/angular/generators';
+import { Tree } from '@nrwl/devkit';
+
+export const createAngularComponentStories = async (
+  tree: Tree,
+  project: string
+): Promise<void> => {
+  await angularStoriesGenerator(tree, {
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/src/utils/create-angular-component-stories.ts:8]: I disagree about it.
Nx recommends using their libraries generators.
https://nx.dev/l/r/generators/workspace-generators

File: packages/ddd/src/generators/init/schema.json [hunk 4]
```diff
@@ -0,0 +1,8 @@
+{
+  "$schema": "https://json-schema.org/schema",
+  "cli": "nx",
+  "$id": "init",
+  "title": "Add ESLint rules for enforce-module-boundaries",
+  "type": "object",
+  "properties": {}
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/src/generators/init/schema.json:7]: Currently, no options are needed here.

File: packages/ddd/src/ddd/ddd-structure.ts [hunk 5]
```diff
@@ -0,0 +1,150 @@
...
+  }
+
+  get depConstraints(): DepConstraint[] {
+    const domainName = this.domainName;
+    if (domainName === this.DOMAIN_SHARED) {
+      return [];
+    }
+    return [
+      {
+        sourceTag: `scope:${domainName}`,
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/src/ddd/ddd-structure.ts:71]: I want to keep `scope` because nx recommends using this word in their book.

File: packages/ddd/README.md [hunk 6]
```diff
@@ -1,7 +1,157 @@
...
+## Basic Usage
+
+### Init
+
+Add ESLint rules for enforce-module-boundaries.
+
+This generator need to be run only once.
+
+```shell
+nx g @e-square/nx-ddd:init
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/README.md:92]: Yea you're right it will be good.
Maybe I will add it in the future.

File: packages/ddd/README.md [hunk 7]
```diff
@@ -1,7 +1,157 @@
...
+npm i -D @nrwl/angular
+```
+
+If you are you going to use React libraries.
+
+```shell
+npm i -D @nrwl/react
+```
+
+**NOTE**: If you are going to use storybook you need to create storybook configuration for the component stories, check out this [video](https://www.youtube.com/watch?v=sFpqyjT7u4s&ab_channel=Nrwl-NarwhalTechnologiesInc.).
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/README.md:81]: To let the people know if they want to use a storybook they need to configure because I generate stories to the components.

File: packages/ddd/src/utils/create-angular-component-stories.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/src/utils/create-angular-component-stories.ts:8]: Why do you add stories? I feel like we need to keep things simple and not combine multiple features into one. its a nice thing to have that if you create a ui library you generate stories for them but what if the user doesn't want it or works with storybook? you kinda force them into it. please remove this part, it should remain up to the users to add it to the lib of their choice

File: packages/ddd/src/utils/create-angular-component.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/src/utils/create-angular-component.ts:18]: disagree about what? I wrote that what you did here is valid because its like exec, you dont import the schematic and it doesn't have to be installed in this project, so that is a good thing.
btw what happens if you dont have the package installed and you use it? does it throw a formatted error?

File: packages/ddd/README.md (hunk 7 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/README.md:81]: continuing the discussion in this comment: https://github.com/e-square-io/nx-plugins/pull/6/files#r739985741

to conclude I believe this should be removed, you're coupling 2 schematics and forcing the user to use storybook

File: packages/ddd/README.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/README.md:92]: i'll open an issue for that

File: packages/ddd/project.json [hunk 8]
```diff
@@ -29,9 +28,15 @@
...
+          "LICENSE",
           {
             "input": "./packages/ddd/src",
-            "glob": "**/*.!(ts)",
+            "glob": "**/!(*.ts)",
+            "output": "./src"
+          },
+          {
+            "input": "./packages/ddd/src",
+            "glob": "**/*.d.ts",
```
Comment (ronnetzer) [inline packages/ddd/project.json:39]: why is this added?

File: packages/ddd/src/generators/init/schema.json (hunk 4 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/src/generators/init/schema.json:7]: why? please be more descriptive, just to remind you of the way of setting global options for schematics, like for angular component to generate only onPush components:
in angular/workspace.json you have schematics part e.g
```JSON
{
  ...,
  "schematics": {
    "@nrwl/angular:component": {
      "style": "scss",
      "changeDetection": "OnPush"
    }
  }
}
```

so that's something that can be done in you init, to set the name of the shared domain, to set defaults or whatever

File: packages/ddd/src/generators/react/generator.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/src/generators/react/generator.ts:3]: so use the wrap thing you did for generating an angular component.
Or, don't instruct your users to install @nrwl/angular or react as you include it internally
Or, move the deps to peerDeps so when users install this lib, it'll check check their deps to see if these installed

**Review by ronnetzer — CHANGES_REQUESTED** [review]
I was sure that I wrote it somewhere, but why in the first place you have an angular/react schematics (framework specific schematics) and not "library"/"domain"/"etc" schematics.

it will be a lot nice to have something like ` nx g @e-square/nx-ddd:library <name> --framework angular (or -f angular) [rest of arguments]`

And will also save you a lot of duplicate code and tests

File: packages/ddd/src/ddd/ddd-structure.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (ronnetzer) [reply packages/ddd/src/ddd/ddd-structure.ts:71]: I understand their recommendation as a general thing, but this is a specific plugin, you dont want it to collide with something else that is unrelated to the plugin, and also you want users to be able to identify what comes from you plugin with ease. Also what if user already has scopes defined and he uses them with some other different logic and grouping? as this is a general thing that nx recommends its something that has high chances of happening.

Maybe we can settle for making this configurable by the user. The default will be `scope` and if he wants to use something else he should be able to do so.

File: packages/ddd/README.md [hunk 9]
```diff
@@ -1,7 +1,120 @@
...
+
+### UI Libraries
+
+A UI library contains only presentational components (also called "dumb" components).
+
+### Data-Access Libraries
+
+A data-access library contains code for interacting with a back-end system. It also includes all the code related to state management.
+
+- `+state` - for state files
```
Comment (ronnetzer) [inline packages/ddd/README.md:54]: what about the `+`?

**Review by ronnetzer — CHANGES_REQUESTED** [review]

File: packages/ddd/README.md [hunk 10]
```diff
@@ -1,7 +1,120 @@
...
+| ------------------- | ----- | -------------------------------------------------------- | ------------- | ----------------------------------------------------------------------------------------------------------------------------------------- |
+| `framework`         | `f`   | 'angular' &#124; 'react'                                 | ''            | The framework to be used for library generation                                                                                           |
+| `type`              | `t`   | 'feature' &#124; 'ui' &#124; 'data-access' &#124; 'util' | 'data-access' | The library type                                                                                                                          |
+| `name`              |       | string                                                   | ''            | The name of the library                                                                                                                   |
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/README.md:112]: It can be provided as the first argument, argv[0] or --name
I think it should be both because of nx console plugin
@ronnetzer

File: packages/ddd/src/generators/init/generator.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+          ...normalizeDDDLibraryGlobalConfiguration(options),
+        },
+      },
+    }
+  );
+
+  const depConstraintsFilePath: string = joinPathFragments(
+    __dirname,
+    'data',
+    'dep-constraints.json'
```
[the remark this answers is not in the document]
Comment (avivharuzi) [reply packages/ddd/src/generators/init/generator.ts]: Changed to const instead of JSON file.
@ronnetzer

File: packages/ddd/README.md (hunk 9 above)
Comment (avivharuzi) [reply packages/ddd/README.md]: Changed +state to state.
@ronnetzer
